### PR TITLE
feat: add manual redirects

### PR DESCRIPTION
**Pull request by crowlKats**

### CODE REVIEW

**Review by ry — COMMENTED** [review]
It would be good to have a test here...

**Comment by kt3k** [comment]
@crowlKats Is this already safe to land in this state?

File: routes/manual.tsx [hunk 1]
```diff
@@ -324,7 +302,40 @@ export const handler: Handlers<Data> = {
...
+          const slug = `${parentSlug}/${childSlug}`;
+          const name = typeof entry === "string" ? entry : entry.name;
+          tempList.push({
+            path: slug,
+            name,
+          });
+          if (typeof entry === "object" && entry.redirectFrom) {
+            for (const redirect of entry.redirectFrom) {
+              redirectList[redirect] = slug;
+            }
```
Comment (kt3k) [inline routes/manual.tsx:320]: We probably need to set correct `redirectFrom` arrays in toc.json of #400?

**Comment by ry** [comment]
@crowlKats I've added a test stub for you, please add some tests.

**Comment by kt3k** [comment]
I believe we need to create mapping from old pages to new pages and set them in `toc.json` in #400

Is this understanding correct?

for example `npm_specifiers` node becomes:
```
{
  "npm_specifiers": "Using Node.js Packages with npm Specifiers",
  ...
}
```
to:

```
{
  "npm_specifiers": {
    "name": "Using Node.js Packages with npm Specifiers",
    "redirectFrom": ["/node/compatibility_mode"]
  },
  ...
}
```

**Comment by kt3k** [comment]
I wonder how many websites link to manual pages without version numbers..

**Review by kt3k — APPROVED** [review]
The change looks good to me. This adds the feature to set up redirects in TOC by `redirectFrom` property.

File: util/manual_utils_test.ts [hunk 2]
```diff
@@ -41,3 +46,53 @@ Deno.test("get edit link new repo", () => {
...
+      "name": "Help",
+    },
+  };
+  const map = tocGen(toc, "foo");
+  assertEquals(
+    map,
+    new Map([
+      ["foo/basics", "Basics"],
+      ["foo/help", "Help"],
+    ]),
```
Comment (ry) [inline util/manual_utils_test.ts:96]: It should test the redirect...

File: routes/manual.tsx [hunk 3]
```diff
@@ -324,7 +287,18 @@ export const handler: Handlers<Data> = {
         }),
     ]);
 
-    return render!({ tableOfContents, content, version });
+    const { pageList, redirectList } = generateToC(
+      tableOfContents,
+      `/manual@${version}`,
+    );
+
+    const slashPath = "/" + params.path;
```
Comment (bartlomieju) [inline routes/manual.tsx:295]: Seems it can't handle a situation when `params.path` starts with a slash.

File: util/manual_utils_test.ts [hunk 4]
```diff
@@ -41,3 +46,53 @@ Deno.test("get edit link new repo", () => {
...
+          },
+        },
+        "debugging_your_code": "Debugging Your Code",
+      },
+    },
+    "help": {
+      "name": "Help",
+    },
+  };
+  const { redirectList } = generateToC(toc, "foo");
```
Comment (bartlomieju) [inline util/manual_utils_test.ts:93]: Please add some assertions for `pageList` as well

File: routes/manual.tsx (hunk 3 above)
Comment (crowlKats) [reply routes/manual.tsx:295]: it never starts with a slash, so not a concern

File: util/manual_utils_test.ts (hunk 4 above)
Comment (crowlKats) [reply util/manual_utils_test.ts:93]: done

**Review by bartlomieju — APPROVED** [review]
LGTM

**Pull request merged**